std/rb.zig

```diff
@@ -0,0 +1,523 @@
+        var maybe_parent: ?*Node = node.get_parent();
+        var color: Color = undefined;
+        var next: *Node = undefined;
+
+        // This clause is to avoid optionals
+        if (node.left == null and node.right == null) {
+            if (maybe_parent) |parent| {
+                parent.set_child(null, parent.left == node);
+            } else
+                tree.root = null;
+            color = node.get_color();
+            newnode = null;
+        } else {
+            if (node.left == null) {
+                next = node.right.?; // Not both null as per above
+            } else if (node.right == null) {
+                next = node.left.?;  // Not both null as per above
+            } else
+                next = node.right.?.get_first(); // Just checked for null above
+
+            if (maybe_parent) |parent| {
+                parent.set_child(next, parent.left == node);
+            } else
+                tree.root = next;
+
+            if (node.left != null and node.right != null) {
+                if (node.left) |left| {
+                    if (node.right) |right| {
+                        color = next.get_color();
+                        next.set_color(node.get_color());
+
+                        next.left = left;
+                        left.set_parent(next);
+
+                        if (next != right) {
+                            var parent = next.get_parent().?; // Was traversed via child node (right/left)
+                            next.set_parent(node.get_parent());
+
+                            newnode = next.right;
+                            parent.left = node;
+
+                            next.right = right;
+                            right.set_parent(next);
+                        } else {
+                            next.set_parent(maybe_parent);
+                            maybe_parent = next;
+                            newnode = next.right;
+                        }
+                    }
+                }
+            } else {
+                color = node.get_color();
+                newnode = next;
+            }
+        }
+
+        if (newnode) |n|
+            n.set_parent(maybe_parent);
+
+        if (color == Red)
+            return;
+        if (newnode) |n| {
+            n.set_color(Black);
+            return;
+        }
+
+        while (true) {
+            if (node == tree.root)
+                break;
+            // If not root, there must be parent
+            var parent = maybe_parent.?;
+            if (node == parent.left) {
+                var sibling = parent.right.?; // Same number of black nodes.
+                
+                if (sibling.is_red()) {
+                    sibling.set_color(Black);
+                    parent.set_color(Red);
+                    rotate_left(parent, tree);
+                    sibling = parent.right.?; // Just rotated
+                }
+                if ((sibling.left == null or sibling.left.?.is_black()) and
```

Comment:
Here's a way to write this without `.?`:

```zig
if ((if (sibling.left) |n| n.is_black() else true) and
    (if (sibling.right) |n| n.is_black() else true))
{
```

Your way is fine too.

Reply:
Yes, I tried to document every time I had to use .? because it is unsafe